### PR TITLE
MLAS: quantized GEMM update

### PR DESCRIPTION
**Pull request by tracysh**
**Description**: Various updates to the int8_t GEMMs:
1. Add ARM64 udot kernel to take advantage of dot product instructions available in newer cores. Some models run 4x faster than the stock implementation we used before.
2. Refactor the x64 kernels to share common code for AVX2(u8u8/u8s8/avxvnni) vs AVX512(u8u8/u8s8/avx512vnni) to reduce binary size.
3. Extend kernels to support per-column zero points for matrix B. This is not currently wired to an operator.

**Motivation and Context**
Besides the above, the entry point for MlasGemm has been changed to take a struct instead of passing down many parameters. This maps to what was already done inside the GEMM routine and also removes duplicated logic from operators to handle packed vs unpacked cases.

The packing logic has also been reworked and is now available on all builds, so there's no longer a need to do #ifdef logic for packed vs unpacked architectures.

### CODE REVIEW

File: onnxruntime/core/mlas/inc/mlas.h [hunk 1]
```diff
@@ -273,41 +269,29 @@ class MLAS_QGEMM_SCALE_BIAS_OUTPUT_PROCESSOR : public MLAS_QGEMM_OUTPUT_PROCESSO
...
-    const uint8_t* B,
-    size_t ldb,
-    uint8_t offb,
-    bool BIsSigned,
-    int32_t* C,
-    size_t ldc,
-    MLAS_THREADPOOL* ThreadPool,
-    const MLAS_QGEMM_OUTPUT_PROCESSOR* OutputProcessor = nullptr
-    );
+struct MLAS_GEMM_U8X8_PARAMETERS {
```
Comment (yufenglee) [inline onnxruntime/core/mlas/inc/mlas.h:272]: like it. It is much easier to use

File: onnxruntime/contrib_ops/cpu/quantization/dynamic_quantize_lstm.cc [hunk 2]
```diff
@@ -33,7 +28,6 @@ class DynamicQuantizeLSTM : public OpKernel, public LSTMBase {
   bool is_R_signed_;
 };
 
-#ifdef MLAS_SUPPORTS_PACKED_GEMM_U8X8
 Status DynamicQuantizeLSTM::TryPackWeights(const Tensor& weights, PackedWeights& packed_weights, bool& is_packed, bool& is_weight_signed) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cpu/quantization/dynamic_quantize_lstm.cc:31]: Could you also clean up this [https://github.com/microsoft/onnxruntime/blob/87ec1f6208f2bb90b079baf444ad942aecfb56ee/onnxruntime/core/providers/cpu/rnn/rnn_helpers.cc#L289](https://github.com/microsoft/onnxruntime/blob/87ec1f6208f2bb90b079baf444ad942aecfb56ee/onnxruntime/core/providers/cpu/rnn/rnn_helpers.cc#L289)?

Comment (tracysh) [reply onnxruntime/contrib_ops/cpu/quantization/dynamic_quantize_lstm.cc:31]: Are you looking at an older branch here? The file you mention was cleaned up (it wouldn't be able to build with the old function signature).

Comment (yufenglee) [reply onnxruntime/contrib_ops/cpu/quantization/dynamic_quantize_lstm.cc:31]: It was in the file list later. 

---
In reply to: [590637096](https://github.com/microsoft/onnxruntime/pull/6916#discussion_r590637096) [](ancestors = 590637096)

File: onnxruntime/core/mlas/lib/qgemm.cpp [hunk 3]
```diff
@@ -17,6 +17,68 @@ Module Name:
...
+struct MLAS_GEMM_U8X8_DISPATCH {
+    MLAS_GEMM_U8X8_OPERATION* Operation;
+    MLAS_GEMM_U8X8_OPERATION* PackedOperation;
+    MLAS_GEMM_U8X8_COPY_PACKB_ROUTINE* CopyPackBRoutine;
+    size_t PackedK;
+    size_t PackedStrideK;
+};
+
+const MLAS_GEMM_U8X8_DISPATCH*
+MlasGemmU8X8GetDispatch(
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/qgemm.cpp:55]: Nice!

File: onnxruntime/core/mlas/lib/x86_64/QgemmU8X8KernelAvx2.S [hunk 4]
```diff
@@ -0,0 +1,868 @@
...
+
+        mov     rbp,rcx                     # reload row length remaining
+
+.LComputeBlockBy1Loop\@:
+        ComputeBlockU8S8\Isa\() \ColumnCount\(), \RowCount\(), 0, 0
+        add     rdi,4                       # advance matrix A by 1 quad
+.if \RowCount\() > 3
+        add     r8,4                        # advance matrix A plus 3 rows by 1 quad
+.endif
+        add     rsi,64                      # advance matrix B
```
Comment (yufenglee) [inline onnxruntime/core/mlas/lib/x86_64/QgemmU8X8KernelAvx2.S:273]: what if ColumnCount is not 16

**Comment by yufenglee** [comment]
>     // This pattern is repeated (CountK / 8) times.

is it 4?

---
Refers to: onnxruntime/core/mlas/lib/qgemm.cpp:1982 in 7a88a97. [](commit_id = 7a88a977797ee3c7bd1b1c50f990228a0d88375a, deletion_comment = False)

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**